boards/arduino-mkr1000: add initial support (without wifi)

aabadie opened the pull request:
This PR adds a partial support for the [Arduino MKR1000](https://www.arduino.cc/en/Main/ArduinoMKR1000) board.

This board uses the same MCU as the Arduino Zero : a SAMD21G18A. But apparently, one has to keep the bootloader at the beginning of the flash memory : as a consequence I added a custom linker script.

The debug UART is also not available via the USB port so one needs an FTDI plugged on the RX/TX pins (pins number 13/14) on the board to access it.

Last thing (but not least) : to flash the board via USB, one needs to use Bossa (like for Arduino-due). For the moment, I have to double press the reset button each time I want to flash the board. This is not very convenient and it's not necessary with the Arduino IDE.

Adding an SPI driver for the ATWINC1500 Wifi module will also be required if one wants to add WiFi to this board.

aabadie commented:
@haukepetersen, interested in testing this one ?

aabadie commented:
I'm planning to add the wifi driver soon using a package from [WiFi101 repository](https://github.com/arduino-libraries/WiFi101). But I prefer adding this in a follow-up PR. Are there people interested in testing this PR so we can merge it fast ?

kbumsik commented:
Hi, I came from #6985 . I am currently merging your branch with my work. I have a question before making a PR to you branch, because I have no experience with programming with BOSSA: Why does Linkerscript's flash address start with 0x2000? Can't BOSSA access to system memory, or you want keep arduino bootloader?

aabadie commented:
@kbumsik I merged your PR. Thanks ! I you could give your test result of this PR (flash, ADC, PWM, UART, etc) that would be great and we could merge it fast.
I'm also looking forward your PR with the ATWINC1500 driver :)

kbumsik commented:
@aabadie Thanks for a quick feedback :) I will work on ATWINC1500 driver targeting both MKR1000 and WINC1500 XPlained Pro.
For testing, I can confirm `example/default` UART terminal example, with pin TX: D14, RX: D13, works perfectly for now. I will update more testing result later.

kbumsik commented:
This needs rebase and fix for ADC after #7009.

aabadie commented:
@kbumsik, I merged #7009 and rebased this branch. I fixed a couple of doxygen issues, added the adc feature for this board and changed the number of ADC channels that was wrong. Let me know if it's fine for you.

kbumsik commented:
I tested the ADC of this board after rebasing and conversion on A0-A6 works well :)

kbumsik commented on boards/arduino-mkr1000/include/periph_conf.h:
```diff
@@ -0,0 +1,273 @@
+};
+
+#define ADC_0_CHANNELS                     (7U)
+#define ADC_NUMOF                          (1U)
```
What about just `#define ADC_NUMOF    ADC_0_CHANNELS`? Firstly all MCUs in the [SAMD family](http://www.atmel.com/products/microcontrollers/arm/sam-d.aspx) have only one ADC, without exception. Also, when we look at the [implementation of the ADC](https://github.com/RIOT-OS/RIOT/blob/master/cpu/samd21/periph/adc.c#L99), you can see it consider the channel as device and the conversion on each channels is done individually. So I think the number of ADC device is practically the same as the number of channels.

aabadie replied:
I pushed the change. And your other remark is totally valid.
Maybe all those defines could be moved in the adc.c file and only the ADC_NUMOF and channels conf struct could be kept in periph_conf.h. This will simplify the configuration and be more consistent with other boards (see [nucleo-f030](https://github.com/RIOT-OS/RIOT/blob/master/boards/nucleo-f030/include/periph_conf.h#L132)). Anyway, I would prefer to have it proposed in a separate PR.

I also noticed that ADC_MAX_CHANNELS is not used in the driver implementation. Is it really needed ?

kbumsik replied:
I don't think ADC_MAX_CHANNELS is neccessary neither. It refers just its specification of the MCU and not useful later. Maybe let us delete it in the later PR, Including samr21/d21-xpro?

aabadie replied:
yes !

kaspar030 commented:
Seems like this PR casues the static tests of murdock2 to time-out.

aabadie commented:
> Seems like this PR casues the static tests of murdock2 to time-out.

Indeed, no idea what could cause this problem.

kbumsik commented:
Test update: I tested I2C and SPI pins using `test/driver_io1_xplained` and `test/driver_sdcard_spi`. Both tests works without problem :)

aabadie commented:
@kaspar030, any idea why Murdock2 fails here ? what is dwqw ?

aabadie commented:
appart from Murdock2 issue, I think we are good to go here. Anyone for an ACK?

kaspar030 commented:
> @kaspar030, any idea why Murdock2 fails here ? what is dwqw ?

"dwqw" is the process executing a murdock2 job. In this case the "static-tests"-job timed out.
Locally executing static tests works fine, but in the riot build container, cppcheck hangs.
Unfortunately I think we need to resolve this before merging...

aabadie commented:
> Unfortunately I think we need to resolve this before merging...

Sure, but what's the difference between Murdock and the local cppcheck ? (I'm running 1.75)

kaspar030 commented:
> Sure, but what's the difference between Murdock and the local cppcheck ? (I'm running 1.75)

I'm getting closer. The build container has cppcheck 1.72. When run without "--quiet", you can see that cppcheck runs through *all* cpp preprocessor combinations, and cppcheck 1.72 seems to be a lot slower than the newer ones...

kaspar030 commented:
Well, I see three options:

- upgrade cppcheck in the build containers (which would go against "we support latest Ubuntu LTS")
- ignore sam0.h for cppcheck
- increase the timeout for static-tests

Anyhow, I ran the static tests of this PR locally, and they succeed. So ACK from my side.

aabadie commented:
@kaspar030, I would go for option 3. Seems to be the more reasonable for me: point 1 break "we support latest Ubuntu LTS" as you said and point 2 breaks full cppcheck support (sam0.h is not a vendor file but RIOT code).

haukepetersen commented on boards/arduino-mkr1000/Makefile.include:
```diff
@@ -0,0 +1,35 @@
+    include $(RIOTMAKE)/tools/jlink.inc.mk
+
+    # set default serial port depending on operating system
+    PORT_LINUX ?= /dev/ttyUSB0
```
This should be ttyACM0 for independent of the programmer, right? So IMHO these two lines should be taken out of this if block.

aabadie replied:
What do you mean with `independent of the programmer` ?
The board USB port is /dev/ttyACM0 but I don't know how it behaves if ones uses the JLINK programmer which I didn't test. If @kbumsik set it to /dev/ttyUSB0 there's probably a good reason.

kbumsik replied:
I copied [Makefile from stm32f4 discovery](https://github.com/RIOT-OS/RIOT/blob/master/boards/stm32f4discovery/Makefile.include#L5). Just like the case in stm32f4 discovery, `/dev/ttyUSB0` is for UART serial converter. But you are right - this is independent to the JLink. I added this when I didn't know how `bossac` works.

aabadie replied:
Oh, ok, I understand now (sorry, for the delay...!)

haukepetersen commented on boards/arduino-mkr1000/Makefile.include:
```diff
@@ -0,0 +1,35 @@
+    # when BOSSA is used, use a different flash map
+    # refer https://github.com/shumatech/BOSSA for this programmer
+    export PROGRAMMER = bossa
+    export CPU_MODEL = samd21g18a_mkr1000
```
Why a different CPU model here? If it is because of the different linkerscript, then it would be the cleaner solution to override the linkerscript variable here directly!

aabadie replied:
I was not aware of this variable, I'll change that as you suggest

haukepetersen commented on boards/arduino-mkr1000/dist/openocd.cfg:
```diff
@@ -0,0 +1,7 @@
+source [find interface/cmsis-dap.cfg]
```
Does this board support being flashed using openocd through CMSIS-DAP? If so, this option is not usable as there is no option the board's Makefile.include... So I'd say either remove this file or add the needed options to the Makefile.include

kbumsik replied:
It is not. We can just remove ir.

haukepetersen commented on boards/arduino-mkr1000/include/board.h:
```diff
@@ -0,0 +1,66 @@
+ */
+#define XTIMER              TIMER_0
+#define XTIMER_CHAN         (0)
+/** @} */
```
no need for this block, as these are the default values. Just remove them.

haukepetersen commented on cpu/sam0_common/include/sam0.h:
```diff
@@ -82,7 +82,7 @@ extern "C" {
 #elif defined(CPU_MODEL_SAMD21G17AU)
   #include "vendor/samd21/include/samd21g17au.h"
-#elif defined(CPU_MODEL_SAMD21G18A)
+#elif defined(CPU_MODEL_SAMD21G18A) || defined(CPU_MODEL_SAMD21G18A_MKR1000)
```
as stated above, there should be no need for this varaible.

haukepetersen commented:
About the Murdock issue: `sam0.h` should be ignored in any case, as this is basically vendor header. My proposed solution: #7038. Does this fix the timeout?

haukepetersen commented:
One more question: I was not able to flash the board, as bossac gives me some GLIB errors:
```
/home/hauke/dev/riot/RIOT/boards/arduino-mkr1000/dist/flash.sh 
/home/hauke/dev/riot/RIOT/boards/arduino-mkr1000/dist/bossac: relocation error: /home/hauke/dev/riot/RIOT/boards/arduino-mkr1000/dist/bossac: symbol _ZNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEE9_M_createERmm, version GLIBCXX_3.4.21 not defined in file libstdc++.so.6 with link time reference
make: *** [flash] Error 127
```
Does anyone have a quick solution to this?

aabadie commented:
> Does anyone have a quick solution to this?

The bossa binary was built on my Linux using gcc 6.2. I probably have a too recent glibc and this is not compatible with yours. I'll rebuild it using an older Ubuntu (16.04)

aabadie commented:
@haukepetersen, except your bossa issue, your other comments are addressed.

haukepetersen commented:
Changes look good. I still can't flash my board (played around with some bossac versions and compiled it myself for a try), but it seems in the end I bricked my board... It doesn't register as /dev/ttyACM0 anymore...

About the bossa issue: I think it would make much sense (as a next step) to include it in form of a Makefile as part of the RIOT tools folder and get rid of the binaries for the arduino sam based boards after all. This way, users of these boards can build their bossac themselves and we get around shipping binaries (which was always sub-optimal). What do you think?

kbumsik commented:
@haukepetersen , I am not sure it is your case, once you flash the MKR1000, it normally won't go to the bootloader mode so you won't see it as tty device. If you want to flash it again, you may double press the reset button and it will go to the bootloader mode.

haukepetersen commented:
Sounds like a plan :-) Will give it a try...

aabadie commented:
> About the bossa issue: I think it would make much sense (as a next step) to include it in form of a Makefile as part of the RIOT tools folder

+1 good idea

aabadie commented:
@haukepetersen, regarding the "bricked" board, as @kbumsik said, just double press the reset button before flashing. I initially added a [comment about that in flash script](https://github.com/RIOT-OS/RIOT/pull/6666/files#diff-ef55d9b7e9b14cd7f867d0f6e13461bcR3)
Thinking of it, the stty line is probably useless (originally copy pasted from arduino-due)

haukepetersen commented on boards/arduino-mkr1000/Makefile.include:
```diff
@@ -0,0 +1,28 @@
+    # when BOSSA is used (default), use a different flash map
+    # refer https://github.com/shumatech/BOSSA for this programmer
+    export PROGRAMMER = bossa
+    export LINKER_SCRIPT ?= $(RIOTCPU)/$(CPU)/ldscripts/$(CPU_MODEL)_mkr1000.ld
```
The path to the linkerscript must be `$(RIOTCPU)/sam0/ldscripts/$(CPU_MODEL)_mkr1000.ld`

aabadie replied:
indeed, missed that change

aabadie replied:
fixed

haukepetersen commented:
Looks good to me now, though the issue with the `bossa` tool remains. We should simply include a global build script/Makefile in RIOT to build that tool - and then use that for the `mkr1000`, the `due`, and `udoo`. Maybe I find time to look into this... But until then, I am ok with merging this PR -> please squash.

aabadie commented:
squashed and rebased.

haukepetersen commented:
Murdock is happy -> let's go!
